### PR TITLE
FEATURE: Improve Node and Node interfaces rewrites

### LINKED ISSUE #57: Migrate `Node::getNodeType`
Following code should be migrated:

For fusion:

- [x] name fetching

```diff
- ${node.nodeType.name}
+ ${q(node).nodeTypeName()} 
```

- [x] nodetype fetching

```diff
- ${node.nodeType}
+ ${Neos.Node.getNodeType(node)}
```

- [ ] For the `Node::getNodeType` removal (https://github.com/neos/neos-development-collection/pull/5021)

```diff
- $node->getNodeTypeName();
+ $node->nodeTypeName;
```

```diff
- $node->getNodeType()->getName();
+ $node->nodeTypeName->value;
```

```diff
- $nodeType->getName();
+ $nodeType->name->value;
```

(existing [migration](https://github.com/neos/rector/blob/4843c45f5bd901bd63aad0d1d0367df394e2cd1e/config/set/contentrepository-90.php#L145))

```diff
- $node->getNodeType();
+ $nodeTypeManger = ...;
+ $nodeTypeManager->getNodeType($node->nodeTypeName);
```

- [ ]  For the `NodeType::tetheredNodeTypeDefinitions` change (https://github.com/neos/neos-development-collection/pull/4906)

These rector migration would need to be adjusted so they dont pass the `$nodeType` to the methods, but `$nodeType->name`:
- NodeTypeGetAutoCreatedChildNodesRector
- NodeTypeGetTypeOfAutoCreatedChildNodeRector
- NodeTypeAllowsGrandchildNodeTypeRector

```diff
- $node->getNodeType()->getTypeOfAutoCreatedChildNode(NodeName::fromString('name'));
+ $nodeTypeManager->getTypeOfTetheredNode($node->nodeTypeName, NodeName::fromString('name'));
```

```diff
- foreach ($nodeType->getAutoCreatedChildNodes() as $rawNodeName => $childNodeType) {
-     $nodeName = NodeName::fromString($rawNodeName);
+ foreach ($nodeType->tetheredNodeTypeDefinitions as $tetheredNodeTypeDefinition) {
+     $nodeName = $tetheredNodeTypeDefinition->name;
+     $childNodeType = $nodeTypeManager->getNodeType($tetheredNodeTypeDefinition->nodeTypeName);

- $nodeType->hasAutoCreatedChildNode($nodeName);
+ $nodeType->tetheredNodeTypeDefinitions->contain($nodeName);

- $parentsNodeType->getTypeOfAutoCreatedChildNode($nodeName)->name;
+ $parentsNodeType->tetheredNodeTypeDefinitions->get($nodeName)->nodeTypeName;

- $parentsNodeType->getTypeOfAutoCreatedChildNode($nodeName);
+ $nodeTypeManager->getNodeType($parentsNodeType->tetheredNodeTypeDefinitions->get($nodeName));

- $grandParentsNodeType->allowsGrandchildNodeType($parentNodeName->value, $nodeType);
+ $nodeTypeManager->isNodeTypeAllowedAsChildToTetheredNode($grandParentsNodeType->name, $parentNodeName, $nodeType->name);
```

### PR DESCRIPTION
**Pull request by dlubitz**
Improves the rewrites for node related code in general.

Fixes #57 
Related: #75

### CODE REVIEW

File: src/ContentRepository90/Rules/NodeGetNodeTypeRector.php [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+        if (!$this->isName($node->name, 'getNodeType')) {
+            return null;
+        }
+
+        $this->nodesToAddCollector->addNodesBeforeNode(
+            [
+                self::withTodoComment(
+                    'Make this code aware of multiple Content Repositories.',
+                    self::assign('contentRepository', $this->this_contentRepositoryRegistry_get($this->contentRepositoryId_fromString('default'))),
```
Comment (kitsunet) [inline src/ContentRepository90/Rules/NodeGetNodeTypeRector.php:55]: I guess this needs to be adapted then as well...

**Comment by kitsunet** [comment]
Great work <3
Left a suggestion, but after that IMHO good to go...

File: docs/rules_overview.md [hunk 2]
```diff
@@ -1480,7 +1480,62 @@ return static function (RectorConfig $rectorConfig): void {
...
+
+ use Neos\Rector\ContentRepository90\Legacy\NodeLegacyStub;
+
+ class SomeClass
+ {
+     public function run(NodeLegacyStub $node)
+     {
+-        $nodeType = $node->getNodeType();
++        // TODO 9.0 migration: Make this code aware of multiple Content Repositories.
++        $contentRepository = $this->contentRepositoryRegistry->get(\Neos\ContentRepository\Core\SharedModel\ContentRepository\ContentRepositoryId::fromString('default'));
```
[the remark this answers is not in the document]
Comment (dlubitz) [reply docs/rules_overview.md:1537]: You are absolutly right. The content repository injection was implemented before the node had the `contentRepositoryId`. Now we can make this more accurate.

[the remark this answers is not in the document]
Comment (dlubitz) [reply docs/rules_overview.md:1537]: But... not in this case :/ 
The line we are rewriting doesn't have the node, just the nodeName.

```diff
- $nodeTypeName = $nodeType->getName();
+ $contentRepository = $this->contentRepositoryRegistry->get(\Neos\ContentRepository\Core\SharedModel\ContentRepository\ContentRepositoryId::fromString('default'));
+ $nodeType = $contentRepository->getNodeTypeManager()->getNodeType($node->nodeTypeName);
```

But I'll check this for other places.

File: src/ContentRepository90/Rules/NodeGetNodeTypeRector.php (hunk 1 above)
Comment (dlubitz) [reply src/ContentRepository90/Rules/NodeGetNodeTypeRector.php]: Done

File: docs/rules_overview.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (dlubitz) [reply docs/rules_overview.md:1537]: Doube checked, where we can use $node->contentRepositoryId directly and replaced the palaces.

File: tests/Sets/ContentRepository90/Fixture/Node/node-properties.php.inc [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+{
+    public function node(\Neos\ContentRepository\Core\Projection\ContentGraph\Node $node)
+    {
+        $value = $node->getProperty('foo');
+        $bool = $node->hasProperty('foo');
+        // TODO 9.0 migration: !! Node::setProperty() is not supported by the new CR. Use the "SetNodeProperties" command to change property values.
+
+
+        $node->setProperty('foo', 'bar');
+        // TODO 9.0 migration: !! Node::setProperty() is not supported by the new CR. Use the "SetNodeProperties" command to remove a property values.
```
Comment (mhsdesign) [inline tests/Sets/ContentRepository90/Fixture/Node/node-properties.php.inc:38]:
```suggestion
        // TODO 9.0 migration: !! Node::removeProperty() is not supported by the new CR. Use the "SetNodeProperties" command to remove a property values.
```

File: src/ContentRepository90/Legacy/NodeLegacyStub.php [hunk 4]
```diff
@@ -21,4 +24,30 @@ public function getNodeType(): NodeType {
...
+    public function findParentNode(): NodeLegacyStub {
+        return new NodeLegacyStub();
+    }
+
+    public function findNamedChildNode(NodeName $nodeName): NodeLegacyStub
+    {
+        return new NodeLegacyStub();
+    }
+
+    public function getNode(): NodeLegacyStub {
```
Comment (mhsdesign) [inline src/ContentRepository90/Legacy/NodeLegacyStub.php:37]: so this thing contains the signature of the original node interface and some of the "forward compatible" changes of the traversable thing ... but this is just a helper during running the migration right? These methods like all the `find*` and `get*` would be the todo to be migrated right?

Comment (dlubitz) [reply src/ContentRepository90/Legacy/NodeLegacyStub.php:37]: Yes, correct. This is just to enable rector to determine the return type of methods, which not exists anymore.

File: tests/Sets/ContentRepository90/Fixture/Node/node-properties.php.inc (hunk 3 above)
Comment (dlubitz) [reply tests/Sets/ContentRepository90/Fixture/Node/node-properties.php.inc:38]: Yeah right. Copy-Paste 🙈

Comment (dlubitz) [reply tests/Sets/ContentRepository90/Fixture/Node/node-properties.php.inc]: Is fixed. Thank you

**Pull request merged**